wgcfg/parser.go

```diff
@@ -0,0 +1,397 @@
+		if equals < 0 {
+			return nil, &ParseError{"Invalid config key is missing an equals separator", line}
+		}
+		key, val := strings.TrimSpace(lineLower[:equals]), strings.TrimSpace(line[equals+1:])
+		if len(val) == 0 {
+			return nil, &ParseError{"Key must have a value", line}
+		}
+		if parserState == inInterfaceSection {
+			switch key {
+			case "privatekey":
+				k, err := ParseKey(val)
+				if err != nil {
+					return nil, err
+				}
+				conf.PrivateKey = PrivateKey(*k)
+				sawPrivateKey = true
+			case "listenport":
+				p, err := parsePort(val)
+				if err != nil {
+					return nil, err
+				}
+				conf.ListenPort = p
+			case "mtu":
+				m, err := parseMTU(val)
+				if err != nil {
+					return nil, err
+				}
+				conf.MTU = m
+			case "address":
+				addresses, err := splitList(val)
+				if err != nil {
+					return nil, err
+				}
+				for _, address := range addresses {
+					a, err := ParseCIDR(address)
+					if err != nil {
+						return nil, err
+					}
+					conf.Addresses = append(conf.Addresses, a)
+				}
+			case "dns":
+				addresses, err := splitList(val)
+				if err != nil {
+					return nil, err
+				}
+				for _, address := range addresses {
+					a, ok := ParseIP(address)
+					if !ok {
+						return nil, &ParseError{"Invalid IP address", address}
+					}
+					conf.DNS = append(conf.DNS, a)
+				}
+			default:
+				return nil, &ParseError{"Invalid key for [Interface] section", key}
+			}
+		} else if parserState == inPeerSection {
+			switch key {
+			case "publickey":
+				k, err := ParseKey(val)
+				if err != nil {
+					return nil, err
+				}
+				peer.PublicKey = *k
+			case "presharedkey":
+				k, err := ParseKey(val)
+				if err != nil {
+					return nil, err
+				}
+				peer.PresharedKey = SymmetricKey(*k)
+			case "allowedips":
+				addresses, err := splitList(val)
+				if err != nil {
+					return nil, err
+				}
+				for _, address := range addresses {
+					a, err := ParseCIDR(address)
+					if err != nil {
+						return nil, err
+					}
+					peer.AllowedIPs = append(peer.AllowedIPs, a)
+				}
+			case "persistentkeepalive":
+				p, err := parsePersistentKeepalive(val)
+				if err != nil {
+					return nil, err
+				}
+				peer.PersistentKeepalive = p
+			case "endpoint":
+				eps, err := parseEndpoints(val)
+				if err != nil {
+					return nil, err
+				}
+				peer.Endpoints = eps
+			default:
+				return nil, &ParseError{"Invalid key for [Peer] section", key}
+			}
+		}
+	}
+	conf.maybeAddPeer(peer)
+
+	if !sawPrivateKey {
+		return nil, &ParseError{"An interface must have a private key", "[none specified]"}
+	}
+	for _, p := range conf.Peers {
+		if p.PublicKey.IsZero() {
+			return nil, &ParseError{"All peers must have public keys", "[none specified]"}
+		}
+	}
+
+	return &conf, nil
+}
+
+// TODO(apenwarr): This is incompatibe with current Device.IpcSetOperation.
+//  It duplicates all the parser stuff in there, but is missing some
+//  keywords. Nothing useful seems to need it anymore.
```

Comment:
If it's broken and nothing uses it, why is this included? Seems like the TODO can be addressed immediately.